src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -1413,7 +1415,8 @@ private void OnHeaderCore(HeaderType headerType, int? staticTableIndex, ReadOnly
         // https://tools.ietf.org/html/rfc7540#section-6.5.2
         // "The value is based on the uncompressed size of header fields, including the length of the name and value in octets plus an overhead of 32 octets for each header field.";
         _totalParsedHeaderSize += HeaderField.RfcOverhead + name.Length + value.Length;
-        if (_totalParsedHeaderSize > _context.ServiceContext.ServerOptions.Limits.MaxRequestHeadersTotalSize)
+        // Allow a 2x grace before aborting the connection. We'll check the size limit again later where we can send a 431.
+        if (_totalParsedHeaderSize > _context.ServiceContext.ServerOptions.Limits.MaxRequestHeadersTotalSize * 2)
```

Comment:
Is this check now duplicated? It's getting tested once in CheckRequestHeadersCountLimit when the header request count increases and again in this method.

Reply:
No, those are different limits. This one is the total length of headers, where the other is the total count of headers.

Reply:
This is checking `MaxRequestHeadersTotalSize` not `MaxRequestHeaderCount`.